### PR TITLE
Fix fails to force_ssl_redirection if session_store is disabled

### PR DESCRIPTION
**Pull request by supercaracal**
#19679

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Could you write a test for it?

**Comment by supercaracal** [comment]
yes sir

**Comment by supercaracal** [comment]
:cry: :bug:

**Comment by supercaracal** [comment]
:cry:

File: actionpack/lib/action_controller/test_case.rb [hunk 1]
```diff
@@ -659,7 +659,9 @@ def process(action, *args)
         @request.assign_parameters(@routes, controller_class_name, action.to_s, parameters)
 
         @request.session.update(session) if session
-        @request.flash.update(flash || {})
+
+        is_request_flash_enabled = @request.respond_to?(:flash)
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/test_case.rb]: Why this change was needed?

**Comment by supercaracal** [comment]
Beacause request reciever has no method "flash".
Only when flash middleware is disabled.

File: actionpack/lib/action_controller/test_case.rb (hunk 1 above)
Comment (rafaelfranca) [inline actionpack/lib/action_controller/test_case.rb]: I mean, this change on the test.

**Comment by supercaracal** [comment]
I'm sorry. Because test is failed by same error on this step.

File: actionpack/lib/action_controller/test_case.rb (hunk 1 above)
Comment (supercaracal) [inline actionpack/lib/action_controller/test_case.rb]: Because request reciever has no method "flash".
Only when flash middleware is disabled.

Comment (supercaracal) [inline actionpack/lib/action_controller/test_case.rb]: I'm sorry. Because test is failed by same error on this step.

Comment (supercaracal) [inline actionpack/lib/action_controller/test_case.rb]: Assignment to extra local variables?

Comment (rafaelfranca) [inline actionpack/lib/action_controller/test_case.rb]: Ah right. Maybe we should move the `enable_flash` call to an `ensure` block at `test_banana_redirects_to_https_if_not_https_and_flash_middleware_is_disabled` and maybe this change is not needed anymore

Comment (supercaracal) [inline actionpack/lib/action_controller/test_case.rb]:
```
kasuga@NPC-0284:/media/kasuga/AREA51/home/vcs/rails$ ruby -Iactionpack/test actionpack/test/controller/force_ssl_test.rb 
Run options: --seed 41545

# Running:

........................E.

Finished in 0.155286s, 167.4329 runs/s, 373.5041 assertions/s.

  1) Error:
RedirectToSSLTest#test_banana_redirects_to_https_if_not_https_and_flash_middleware_is_disabled:
NoMethodError: undefined method `flash' for #<ActionController::TestRequest:0x007fddd0b00dc8>
    /media/kasuga/AREA51/home/vcs/rails/actionpack/lib/action_controller/test_case.rb:662:in `process'
    /media/kasuga/AREA51/home/vcs/rails/actionpack/lib/action_controller/test_case.rb:64:in `process'
    /media/kasuga/AREA51/home/vcs/rails/actionpack/lib/action_controller/test_case.rb:760:in `process_with_kwargs'
    /media/kasuga/AREA51/home/vcs/rails/actionpack/lib/action_controller/test_case.rb:515:in `get'
    actionpack/test/controller/force_ssl_test.rb:327:in `test_banana_redirects_to_https_if_not_https_and_flash_middleware_is_disabled'

26 runs, 58 assertions, 0 failures, 1 errors, 0 skips
```

:sob:

Comment (supercaracal) [inline actionpack/lib/action_controller/test_case.rb]: I moved the `enable_flash` call to an ensure block. Thank you bery much.
But, This `@request.respond_to?(:flash)` change was needed.

**Comment by supercaracal** [comment]
ci rebuild plz

**Comment by rafaelfranca** [comment]
Awesome work

**Comment by rafaelfranca** [comment]
Backported as c688b07

**Comment by supercaracal** [comment]
:congratulations:

**Comment by spastorino** [comment]
I don't think this change is OK and is breaking Rails API.

What if your controller does not include AC::Flash? we need to check for respond_to?(:flash) and it's the right thing to do there.

I'm reverting this until we find a proper fix for it. @supercaracal can you provide a sample app that reproduces the issue you're having? because I also feel that the provided tests are wrong.

https://github.com/rails/rails/commit/e085a7ddb879430133c68ccfea1f2be5cccdcfa3 and https://github.com/rails/rails/commit/4a2c7c6e7ec88759ae5ac2489b093a2b8330809c

**Comment by supercaracal** [comment]
@spastorino I'm sorry. My app config is https://github.com/rails/rails/issues/19679

We use Redis Cluster Session Store. w/original implements

**Comment by supercaracal** [comment]
I'm not good at English. I'm sorry.

**Comment by spastorino** [comment]
@supercaracal don't worry about english. My english is not good neither :).

Let me check if I understand correct ... so basically you want flash disabled in your app?.

If so, you need to remove the middleware and exclude the module in AC.
Seems like you're not excluding Flash in AC.
If you inherit from AC::Base you end having Flash, in order to avoid Flash you need to do something like this ...

```
   class ApplicationController
     ActionController::Base.without_modules(:Flash).each do |left|
       include left
     end
   end
```

**Comment by rafaelfranca** [comment]
Hmm, makes sense. But I wish we have an easy way to disable flash entirely.
Having to remember to remove a middleware and a module for your default
controller is too much work.

**Comment by supercaracal** [comment]
Thank you very much. @spastorino @rafaelfranca 

The `Flash` is useful feature. I will use it in future.

I think that the `Flash` should be pluggable module.
But, We can not the `force_ssl` without the `Flash` feature. It seems like a bug.

And yet, I have no an easy implementation to the `Rails Session Store` w/ the `Redis Cluster` by major gems. So I was to disabled the `Rails Session Store`.

**Comment by rafaelfranca** [comment]
:confounded: tricky. We definitely should decouple these things. @supercaracal could you create an example application doing exactly what you are doing in your application so we can use as base to decouple these modules and create a better API for removing some features like Flash or Session Store?

**Comment by supercaracal** [comment]
Let me try. Just a moment please. @rafaelfranca

**Comment by supercaracal** [comment]
@rafaelfranca @spastorino https://github.com/supercaracal/session_store_less_sample I apologise for the delay in replying to you.

**Comment by supercaracal** [comment]
is related? https://github.com/rails/rails/pull/19832

**Comment by rafaelfranca** [comment]
@supercaracal no problem. Thank you so much for writing this.

**Comment by rafaelfranca** [comment]
I'll play with this application and see how we can make easier to disable this feature.

**Comment by prathamesh-sonpatki** [comment]
@supercaracal The application link is giving 404 https://github.com/supercaracal/session_store_less_sample. Can you please upload again, thanks!

**Comment by supercaracal** [comment]
@prathamesh-sonpatki @rafaelfranca I'm terribly sorry. I deleted it before. Because Rails has already API Mode. But, I can ready it once again. Just a moment please.

**Comment by supercaracal** [comment]
@prathamesh-sonpatki @rafaelfranca I created it again. https://github.com/supercaracal/session_store_less_sample

**Comment by prathamesh-sonpatki** [comment]
Thanks!

**Comment by prathamesh-sonpatki** [comment]
@rafaelfranca @spastorino @supercaracal 

I looked into this issue and found out the real problem. The issue is we include Flash middleware only if session_store is not [disabled](https://github.com/rails/rails/blob/6d6249b1c1abda4f62fafcc42a7ece570c8da7e9/railties/lib/rails/application/configuration.rb#L182-L183) -- https://github.com/rails/rails/blob/6d6249b1c1abda4f62fafcc42a7ece570c8da7e9/railties/lib/rails/application/default_middleware_stack.rb#L58-L64

So the request object [does not get flash method](https://github.com/rails/rails/blob/6d6249b1c1abda4f62fafcc42a7ece570c8da7e9/actionpack/lib/action_dispatch/middleware/flash.rb#L294-L296).

The fix in this PR is the right fix to check whether request object responds to flash method or not, because the controller object will always respond to the [flash method](https://github.com/rails/rails/blob/6d6249b1c1abda4f62fafcc42a7ece570c8da7e9/actionpack/lib/action_controller/metal/flash.rb#L9) in a [non-API Rails app](https://github.com/rails/rails/blob/6d6249b1c1abda4f62fafcc42a7ece570c8da7e9/actionpack/lib/action_controller/base.rb#L223) but whether request object responds to flash method or not is decided based on whether Flash middleware gets included or not.

--------------

The issue faced by @spastorino for API apps was real when this PR got merged: https://github.com/rails/rails/blob/4a2c7c6e7ec88759ae5ac2489b093a2b8330809c/railties/lib/rails/application/default_middleware_stack.rb#L47 but later Santiago fixed it by checking first that `config.api_mode` is set or not --https://github.com/rails/rails/commit/135c059d6f0664991f158d358a4312e550ead7a6

So now this PR change does not affect API only apps and can be merged safely.

---------------

Also this issue only arises when session store is disabled, not when a different session store is used. As Flash depends on session store, Flash not working after disabling session store seems expected outcome to me. May be we need to document it clearly?

**Comment by spastorino** [comment]
So can you guys provide a new PR and ping me?. Thanks.

**Comment by prathamesh-sonpatki** [comment]
:+1: @supercaracal Can you send same PR :smile:

**Comment by supercaracal** [comment]
I will do it. :fire:

**Comment by supercaracal** [comment]
@spastorino @prathamesh-sonpatki https://github.com/rails/rails/pull/27140
